aiokafka/protocol/api.py

```diff
@@ -59,32 +64,103 @@ class ResponseHeader_v1(Struct):
+        the API version
+    """
+
+    API_KEY: ClassVar[int]
+    CLASSES: ClassVar[list[type[RequestStruct]]]
+    ALLOW_UNKNOWN_API_VERSION: ClassVar[bool] = False
+
+    def __init_subclass__(cls) -> None:
+        super().__init_subclass__()
+        if cls.API_KEY is None or cls.CLASSES is None:
+            raise TypeError(
+                f"{cls.__name__} must define class attributes 'API_KEY' and 'CLASSES"
+            )
 
-    @property
     @abc.abstractmethod
-    def API_KEY(self) -> int:
-        """Integer identifier for api request"""
+    def build(self, request_struct_class: type[RequestStruct]) -> RequestStruct:
+        pass
+
+    def prepare(self, versions: dict[int, tuple[int, int]]) -> RequestStruct:
+        api_key = self.API_KEY
+        if api_key not in versions:
+            if self.ALLOW_UNKNOWN_API_VERSION:
+                return self.build(self.CLASSES[0])
+            raise IncompatibleBrokerVersion(
+                f"{self.__class__.__name__} cannot be used "
+                "if the API version is unknown"
+            )
 
-    @property
-    @abc.abstractmethod
-    def API_VERSION(self) -> int:
-        """Integer of api request version"""
+        min_version, max_version = versions[api_key]
+        for req_class in reversed(self.CLASSES):
+            if min_version <= req_class.API_VERSION <= max_version:
+                return self.build(req_class)
 
-    @property
-    @abc.abstractmethod
-    def RESPONSE_TYPE(self) -> type[Response]:
-        """The Response class associated with the api request"""
+        raise NotImplementedError(
+            f"Support for {self.__class__.__name__} v{min_version} "
+            "has not yet been added"
+        )
 
-    @property
-    @abc.abstractmethod
-    def SCHEMA(self) -> Schema:
-        """An instance of Schema() representing the request structure"""
 
-    def expect_response(self) -> bool:
-        """Override this method if an api request does not always generate a response"""
-        return True
+class RequestStruct(Struct, metaclass=abc.ABCMeta):
+    """
+    Base structure for API requests.
+
+    Attributes
+    ----------
+    FLEXIBLE_VERSION : bool
+        Use request header with flexible tags
+    API_KEY : int
+        The unique API key identifying the request.
+    API_VERSION : int
+        Which API version the RequestStruct class is.
+    RESPONSE_TYPE : type[Response]
+        Class used to parse the response.
+    SCHEMA : Schema
+        An instance of Schema() representing the request structure.
+    """
+
+    FLEXIBLE_VERSION: ClassVar[bool] = False
+    API_KEY: ClassVar[int]
+    API_VERSION: ClassVar[int]
+    RESPONSE_TYPE: ClassVar[type[Response]]
+    SCHEMA: ClassVar[Schema]
+
+    def __init_subclass__(cls) -> None:
+        super().__init_subclass__()
+        if (
+            cls.API_KEY is None
+            or cls.API_VERSION is None
+            or cls.RESPONSE_TYPE is None
+            or cls.SCHEMA is None
```

Comment:
If we forget to define some of them, they still won't be None. Probably we don't need those checks at all?

Reply:
There is a test in test_requests.py and it will raise an AttributeError. Maybe I could do a better testing to raise a single type of error (like "the attribute need to be define + being not None and the proper type" ?)